docs/src/best-practices-js.md

```diff
@@ -0,0 +1,200 @@
+
+Traces are set to run on CI on the first retry of a failed test. We don't recommend setting this to 'on' so that traces are run on every test as it's very performance heavy. However you can run a trace locally when developing with the `--trace` flag.
+
+```js
+npx playwright test --trace on
+```
+### Use Playwright's Tooling
+
+Playwright comes with a range of tooling to help you write tests. 
+- The [VS Code extension](./getting-started-vscode.md) gives you a great developer experience when writing, running and debugging tests. 
+- The [test generator](./codegen.md) can generate tests and pick locators for you.
+- The [trace viewer](./trace-viewer.md) gives you a full trace of your tests as a local PWA that can easily be shared. With the trace viewer you can view the timeline, inspect DOM snapshots for each action, view network requests and more.
+- [Typescript](./test-typescript) in Playwright works out of the box and gives you better IDE integrations. Your IDE will show you everything you can do and highlight when you do something wrong. No TypeScript experience is needed and it is not necessary for your code to be in TypeScript, all you need to do is create your tests with a `.ts` extension.
+
+### Test across all browsers
+
+Playwright makes it easy to test your site across all [browsers](./test-configuration#multiple-browsers) no matter what platform you are on. Testing across all browsers ensures your app works for all users. In your config file you can set up projects adding the name and which browser or device to use.
+
+```js tab=js-js
+// playwright.config.js
+// @ts-check
+const { devices } = require('@playwright/test');
+
+/** @type {import('@playwright/test').PlaywrightTestConfig} */
+const config = {
+  projects: [
+    {
+      name: 'chromium',
+      use: { ...devices['Desktop Chrome'] },
+    },
+    {
+      name: 'firefox',
+      use: { ...devices['Desktop Firefox'] },
+    },
+    {
+      name: 'webkit',
+      use: { ...devices['Desktop Safari'] },
+    },
+  ],
+};
+
+module.exports = config;
+```
+
+```js tab=js-ts
+// playwright.config.ts
+import { type PlaywrightTestConfig, devices } from '@playwright/test';
+
+const config: PlaywrightTestConfig = {
+  projects: [
+    {
+      name: 'chromium',
+      use: { ...devices['Desktop Chrome'] },
+    },
+    {
+      name: 'firefox',
+      use: { ...devices['Desktop Firefox'] },
+    },
+    {
+      name: 'webkit',
+      use: { ...devices['Desktop Safari'] },
+    },
+  ],
+};
+export default config;
+```
+
+### Mock third-party API responses
+
+If you want to test links to third party API's then you should intercept the route with a mock response. This ensures the link is visible and clickable. Before hitting the link the route gets intercepted and a mock response is returned. Clicking the link results in a new page being opened containing the mock response rather than the actual page. We can then check this has the URL we expect.
```

Comment:
I don't follow the connection between this paragraph of text and the code sample below. The code shows following a normal HTML link it seems. I don't see anything about a "third party API".

Reply:
but its not a normal HTML link. its a link to an external site like GitHub or could be twitter etc so you are hitting their API. So image testing the social links in your menu. you want to make sure the link is clickable so people can follow you. But you dont want to hit GitHubs or Twitters server. Could be called Mock third-party links but that sounds weird.